
# 스프링의 정석
- 2022.06.16(목)
- 강의 목표: Spring의 기본 원리 이해, Spring으로 웹 사이트 구축 & AWS 배포
- 강의 자료: https://github.com/castello/spring_basic
- 강의 진도 
	- Ch0227. 예외처리(1) - 실습
	- Ch0228. 예외처리(2) - 이론

<br>

## 예외처리-실습
- 실습 파일 
	- `error.jsp`: https://github.com/castello/spring_basic/blob/main/ch2/error.jsp 에서 복붙
	- `ExceptionController.java`
	- `GlobalCatcher.java`
	- 확인: http://localhost:8080/jinsook/ex

<br>

```java
//ExceptionController.java
@Controller
public class ExceptionController {
	@RequestMapping("/ex")
	public void main() throws Exception{
		throw new Exception("예외 발생");
	}
}
```
- 예외가 발생했는데, 내부에서 해결하지 않으면 500 에러 발생

```java
//ExceptionController.java
@Controller
public class ExceptionController {
	@RequestMapping("/ex")
	public void main() throws Exception{
		try {
			throw new Exception("예외 발생");
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
```
- 예외 처리 하고자하는 코드 선택 > 우클릭 > Surround withd > try catch block > 404 에러 발생
- Mapping된 URL로 접근 했을 때, 해당하는 view가 없으므로 404 에러 발생  

```java
//ExceptionController.java
@Controller
public class ExceptionController {
	@RequestMapping("/ex")
	public String main() throws Exception{
		try {
			throw new Exception("예외 발생");
		} catch (Exception e) {
			return "error"; // error.jsp로 연결
		}
	}
}
```
- main 메서드에서 jsp를 return하여 이동하도록 지정한다

```java
//ExceptionController.java
@Controller
public class ExceptionController {
	@RequestMapping("/ex")
	public String main() throws Exception{
		try {
			throw new Exception("예외 발생1");
		} catch (Exception e) {
			return "error"; // error.jsp로 연결
		}
	}

	@RequestMapping("/ex2")
	public String main2() throws Exception{
		try {
			throw new Exception("예외 발생2");
		} catch (Exception e) {
			return "error"; // error.jsp로 연결
		}
	}
}
```
```java
@ExceptionHandler(Exception.class)
public String catcher(Exception ex) {
	return "error";
}	

@RequestMapping("/ex")
public String main() throws Exception{
	throw new Exception("예외 발생1");
}	

@RequestMapping("/ex2")
public String main2() throws Exception{
	throw new Exception("예외 발생2");
}
```
- 만약 동일한 try-catch구문이 반복된다면 예외처리를 위한 메서드로 빼내서 사용할 수 있다
- 예외처리를 위한 메서드에 `@ExceptionHandler` 어노테이션을 붙여주면, 매개변수에 있는 예외가 발생한 경우 해당 예외를 처리해준다
- view(error.jsp)에서 어떤 예외가 발생했는지 보여주도록 수정해보자

```java
@ExceptionHandler(Exception.class)
public String catcher(Exception ex, Model m) {
	m.addAttribute("ex", ex);
	return "error";
}	
```
- Controller가 View에 데이터를 전달하려면 Model에 데이터를 담아야한다. Model을 매개변수에 추가한다
- 위와 같이 코드를 수정하면, 발생한 예외에 대한 정보가 Model에 담겨 View에 전달된다
```jsp
<!-- error.jsp -->
<h1>예외가 발생했습니다.</h1>
발생한 예외 : ${ex}<br>
예외 메시지 : ${ex.message}<br>
```
- View에 해당하는 `error.jsp` 코드를 살펴보면, Model에 `ex`라는 key값으로 저장되어 전달한 예외의 정보를 el형식으로 출력하고 있다

```java
// 특정 예외 하나를 처리하는 catcher1 메서드 
@ExceptionHandler(NullPointerException.class)
public String catcher1(NullPointerException ex, Model m) {
	m.addAttribute("ex", ex);
	return "error";
}	

// 여러 예외를 처리하는 catcher2 메서드 (배열 사용)
@ExceptionHandler({RuntimeException.class, FileNotFoundException.class})
public String catcher2(Exception ex, Model m) {
	m.addAttribute("ex", ex);
	return "error";
}	

// 최고 조상인 Exception을 처리하는 catcher3 메서드
/// 만약 catcher1에서도, catcher2에서도 처리하지 못하는 예외의 경우
/// 최고 조상을 처리하는 catcher3 메서드가 처리한다 
@ExceptionHandler(Excpetion.class)
public String catcher1(Excpetion ex, Model m) {
	m.addAttribute("ex", ex);
	return "error";
}	
```
- 여러가지 예외를 동일하게 대응하는 메서드의 경우, `@ExceptionHandler` 어노테이션에 지정하는 예외를 배열에 담아 지정한다
- 만약 대응하는 메서드가 없는 경우를 대비해 최고 조상인 Exception을 처리하는 메서드를 만들어 두는 편이 좋다
- 모든 Controller마다 예외처리를 하는 메서드를 만드는 것은 비효율적이므로 예외처리를 위한 클래스를 만들어 import하여 사용하도록한다

```java
// @ControllerAdvice 					// 모든 패키지에 적용
@ControllerAdvice(com.study.jinsook) 	// 정해진 패키지에만 적용 
public class GlobalCatcher {	
	@ExceptionHandler(Exception.class)
	public String catcher(Exception ex, Model m) {
		m.addAttribute("ex", ex);
		return "error";
	}
	
}
```
- `@ControllerAdvice`를 붙여주게되면 모든 Controller에서 발생한 예외를 catcher 메서드가 처리해주게 된다
	- `@ControllerAdvice`를 적용할 패키지를 지정할 수 있다. 생략하면 전체 패키지에 적용된다
- 해당 Controller 내부에도 예외처리 메서드가 있다면, 더 가까이에 있는 예외처리 메서드가 적용된다 (Global X)